### PR TITLE
Updated 'mingw-w64-arm-none-eabi-gdb' PKGBUILD

### PR DESCRIPTION
**Pull request by devjoa**
PKGBUILD changed to a mash-up between 'mingw-w64-gdb' and
Arch Linux 'arm-none-eabi-gdb'

Fix Python dependency issues, and build dependencies

### CODE REVIEW

**Comment by lazka** [comment]
Not a fan of adding patches without knowing why...

@fauxpark

**Comment by fauxpark** [comment]
Probably the only thing that needs to be changed here is the missing dependencies added, which I was going to do soon...

**Comment by devjoa** [comment]
Sorry, my bad. The commit was incomplete, and I should neither pull a commit with that low of information. To my excuse, I made it in the morning, and I just wanted to continue with my day job (as I need this to make it all work). The error I was trying to fix is reported in https://github.com/msys2/MINGW-packages/issues/6997

The today build for arm-none-eabi-gdb is broken, as I get the following error. 
```
$ arm-none-eabi-gdb --version
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Python path configuration:
  PYTHONHOME = (not set)
  PYTHONPATH = (not set)
  program name = 'D:/a/_temp/msys/msys64/mingw64/bin/python'
  isolated = 0
  environment = 1
  user site = 1
  import site = 1
  sys._base_executable = 'D:/a/_temp/msys/msys64/mingw64/bin/python'
  sys.base_prefix = 'D:/a/_temp/msys/msys64/mingw64'
  sys.base_exec_prefix = 'D:/a/_temp/msys/msys64/mingw64'
  sys.executable = 'D:/a/_temp/msys/msys64/mingw64/bin/python'
  sys.prefix = 'D:/a/_temp/msys/msys64/mingw64'
  sys.exec_prefix = 'D:/a/_temp/msys/msys64/mingw64'
  sys.path = [
    'D:/a/_temp/msys/msys64/mingw64/lib/python38.zip',
    'D:/a/_temp/msys/msys64/mingw64/lib/python3.8',
    'D:/a/_temp/msys/msys64/mingw64/lib/python3.8',
    'D:/a/_temp/msys/msys64/mingw64/lib/lib-dynload',
  ]
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'

Current thread 0x000c4ee8 (most recent call first):
<no Python frame>
```

What I've done is that I've modified the ordinary MSYS2 mingw-gdb PKGBUILD https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-gdb, and changed the target, and used the target options defined for Arch Linux, together with their naming convention..

```
    --target=${_target} 
    --enable-languages=c,c++
```

See my updated commit, with all the missing patches.

**Comment by devjoa** [comment]
....and here is the PKGBUILD differences between the ordinary mingw-gdb

```
$ diff ./mingw-w64-arm-none-eabi-gdb/PKGBUILD ./mingw-w64-gdb/PKGBUILD
2,4c2
< # Maintainer: fauxpark <fauxpark@gmail.com>
< # Contributed by: nanodude
< # Contributed by: devjoa
---
> # Contributor: Ray Donnelly <mingw.android@gmail.com>
7,9c5,6
< _target=arm-none-eabi
< pkgbase=mingw-w64-${_target}-${_realname}
< pkgname="${MINGW_PACKAGE_PREFIX}-${_target}-${_realname}"
---
> pkgbase=mingw-w64-${_realname}
> pkgname="${MINGW_PACKAGE_PREFIX}-${_realname}"
12c9
< pkgdesc="GNU Debugger (${_target})"
---
> pkgdesc="GNU Debugger (mingw-w64)"
16c13
< groups=("${MINGW_PACKAGE_PREFIX}-${_target}-toolchain")
---
> groups=("${MINGW_PACKAGE_PREFIX}-toolchain")
37c34
< validpgpkeys=('F40ADB902B24264AA42E50BF92EDB04BFF325CF3') # Joel Brobecker
---
> validpgpkeys=('F40ADB902B24264AA42E50BF92EDB04BFF325CF3')
74,75c71,72
<   [[ -d ${srcdir}/build-${MINGW_CHOST}-${_target} ]] && rm -rf ${srcdir}/build-${MINGW_CHOST}-${_target}
<   mkdir ${srcdir}/build-${MINGW_CHOST}-${_target} && cd ${srcdir}/build-${MINGW_CHOST}-${_target}
---
>   [[ -d ${srcdir}/build-${MINGW_CHOST} ]] && rm -rf ${srcdir}/build-${MINGW_CHOST}
>   mkdir ${srcdir}/build-${MINGW_CHOST} && cd ${srcdir}/build-${MINGW_CHOST}
95c92
<     --target=${_target} \
---
>     --target=${MINGW_CHOST} \
97c94
<     --enable-languages=c,c++ \
---
>     --enable-targets="i686-w64-mingw32,x86_64-w64-mingw32" \
117,118c114,115
<   cd ${srcdir}/build-${MINGW_CHOST}-${_target}
<   make DESTDIR=${pkgdir} install-gdb
---
>   cd ${srcdir}/build-${MINGW_CHOST}
>   make DESTDIR=${pkgdir} install
121c118
<   rm -rf ${pkgdir}${MINGW_PREFIX}/share/{gdb,man,info}
---
>   rm -rf ${pkgdir}${MINGW_PREFIX}/share/{man,info}
123,124c120,121
<   rm -rf ${pkgdir}${MINGW_PREFIX}/include
<   rm -rf ${pkgdir}${MINGW_PREFIX}/lib
---
>   rm -f ${pkgdir}${MINGW_PREFIX}/include/*.h
>   rm -f ${pkgdir}${MINGW_PREFIX}/lib/*.a

```

**Comment by fauxpark** [comment]
@devjoa are you sure all of these patches are necessary, though? I have a feeling most of them aren't anymore. I'd prefer if you kept the changes to the minimum necessary to get it working, it will probably save the maintainers some headache. Especially things like changing the pkgdesc, now it does not match the rest of the arm-none-eabi toolchain packages.

**Comment by devjoa** [comment]
No, I'm not sure. Don't know what all patches does, but it works for me, and as I mentioned, they are all from mingw-gdb (and I thought they are there for some reason).

However, I don't understand why the two PKGBUILD's should differ so much, as all added patches seems to affect the host, and not the target? I made the patch due to the arm-none-eabi-gdb has been broken for a while, and when the new version didn't work neither I made an attempt to fix the problems, and with this patch I don't have to use 'https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm'.

I won't do any deep analysis why the previous PKBUILD didn't work, because made the changes due to that I knew mingw-gdb worked, and mingw-arm-none-eabi-gdb did not, and the most natural would be that mingw-arm-none-eabi-gdb derives from that one, and I changed the naming convention to the one that i thought was msys2 standard (since msys2 seems to have an Arch Linux approach).  It's dumb to duplicate work, but it's not my intention to offend anyone. 

I'm not offended if you reject my pull request, I just wanted to help the community.

**Comment by devjoa** [comment]
@fauxpark, I've updated the commit with a (hopefully) more descriptive commit comment, and I've removed one of the patches. The others seemed (when reading the patch comments) to affect stability and performance, and thats why I didn't remove them.
I've also restructured the PKBUILD to be more diff friendly. Hope this is up to your standards now.
The enabling of TUI mode may be unwanted, but I think it's pretty neat for fast command-line debug sessions.

You're free to modify my PR if you want to remove/add stuff, I didn't expect a clean commit from the beginning.

File: mingw-w64-arm-none-eabi-gdb/PKGBUILD [hunk 1]
```diff
@@ -8,30 +8,84 @@ _target=arm-none-eabi
...
+makedepends=("${MINGW_PACKAGE_PREFIX}-iconv"
+             "${MINGW_PACKAGE_PREFIX}-xz")
+options=('staticlibs' '!distcc' '!ccache')
 validpgpkeys=('F40ADB902B24264AA42E50BF92EDB04BFF325CF3') # Joel Brobecker
 
 prepare() {
     cd ${srcdir}/${_realname}-${pkgver}
 
-    mkdir ${_realname}-build
+
```
Comment (fauxpark) [inline mingw-w64-arm-none-eabi-gdb/PKGBUILD:54]:
```suggestion
    mkdir ${_realname}-build-${MINGW_CHOST}

```

File: mingw-w64-arm-none-eabi-gdb/PKGBUILD [hunk 2]
```diff
@@ -8,30 +8,84 @@ _target=arm-none-eabi
...
+    patch -p1 -i ${srcdir}/gdb-home-is-userprofile.patch
+
+    # hack! - libiberty configure tests for header files using "$CPP $CPPFLAGS"
+    sed -i "/ac_cpp=/s/\$CPPFLAGS/\$CPPFLAGS -O2/" libiberty/configure
 }
 
 build() {
-    cd ${srcdir}/${_realname}-${pkgver}/${_realname}-build
+    [[ -d ${srcdir}/${_realname}-${pkgver}/${_realname}-build-${MINGW_CHOST} ]] && rm -rf ${srcdir}/${_realname}-${pkgver}/${_realname}-build-${MINGW_CHOST}
+    mkdir ${srcdir}/${_realname}-${pkgver}/${_realname}-build-${MINGW_CHOST} && cd ${srcdir}/${_realname}-${pkgver}/${_realname}-build-${MINGW_CHOST}
```
Comment (fauxpark) [inline mingw-w64-arm-none-eabi-gdb/PKGBUILD:75]:
```suggestion
    cd ${srcdir}/${_realname}-${pkgver}/${_realname}-build-${MINGW_CHOST}
```

File: mingw-w64-arm-none-eabi-gdb/PKGBUILD [hunk 3]
```diff
@@ -8,30 +8,84 @@ _target=arm-none-eabi
...
+    fi
+
+    CPPFLAGS+=" -I${MINGW_PREFIX}/include/ncurses"
+    CFLAGS+=" -I${MINGW_PREFIX}/include/ncurses"
+    CXXFLAGS+=" -I${MINGW_PREFIX}/include/ncurses"
+    LDFLAGS+=" -fstack-protector"
 
     ../configure \
         --build=${MINGW_CHOST} \
+        --host=${MINGW_CHOST} \
```
Comment (fauxpark) [inline mingw-w64-arm-none-eabi-gdb/PKGBUILD:88]:
```suggestion
```

File: mingw-w64-arm-none-eabi-gdb/PKGBUILD [hunk 4]
```diff
@@ -8,30 +8,84 @@ _target=arm-none-eabi
...
+    patch -p1 -i ${srcdir}/gdb-fix-using-gnu-print.patch
+
+    # https://sourceware.org/bugzilla/show_bug.cgi?id=21078
+    patch -p1 -i ${srcdir}/gdb-7.12-dynamic-libs.patch
+
+    patch -p1 -i ${srcdir}/python-configure-path-fixes.patch
+
+    patch -p1 -i ${srcdir}/gdb-fix-tui-with-pdcurses.patch
+    patch -p1 -i ${srcdir}/gdb-9.1-lib-order.patch
+    patch -p1 -i ${srcdir}/gdb-home-is-userprofile.patch
```
Comment (fauxpark) [inline mingw-w64-arm-none-eabi-gdb/PKGBUILD:67]:
```suggestion
    patch -p1 -i ../gdb-perfomance.patch

    patch -p1 -i ../gdb-fix-using-gnu-print.patch

    # https://sourceware.org/bugzilla/show_bug.cgi?id=21078
    patch -p1 -i ../gdb-7.12-dynamic-libs.patch

    patch -p1 -i ../python-configure-path-fixes.patch

    patch -p1 -i ../gdb-fix-tui-with-pdcurses.patch
    patch -p1 -i ../gdb-9.1-lib-order.patch
    patch -p1 -i ../gdb-home-is-userprofile.patch
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by devjoa** [comment]
@fauxpark, I've commited and resolved your proposed changes. All of them seemed fair. :-)

**Comment by lazka** [comment]
If there is a agreement that this is ready, please ping me.

**Comment by fauxpark** [comment]
I'm not 100% sure whether most of the patches are necessary, but if it compiles and works that's a ✔️ from me. (I unfortunately don't have any hardware to fully test it and don't know a whole lot about how to work GDB)

**Comment by lazka** [comment]
thanks. let's see..

**Pull request merged**